Assignment4/PointGrid.py

```diff
@@ -0,0 +1,170 @@
+class Grid:
+
+    def __init__(self, grid):
+        """
+        For grid initialization with a grid with elements.
+        :param grid: list(list(any))
+        :return: None
+        """
+        self._grid = grid
+        self._points = {}
+        self._update_points()
+        self._islands = []
+
+    class Point:
+        def __init__(self, value, coordinates):
+            self.coordinates = coordinates
+            self.value = value
+            self.is_visited = False
+
+        def visit(self):
+            """
+            Set the visit attribute to be True.
+            :return: None
+            """
+            self.is_visited = True
+
+        def __eq__(self, other):
+            """
+            Supports equivalence for the Point. Returns True if the elements self and other are equal, False otherwise.
+            :param other: any
+            :return: bool
+            """
+            if isinstance(other, self.__class__):
+                return self.coordinates == other.coordinates and self.value == other.value and self.is_visited == other.is_visited
+            else:
+                return False
+
+    def find_point(self, point_value):
+        """
+        Return the first NOT VISITED Point in self._points with the given value if exists. None otherwise.
+        :param point_value: any
+        :return: Point or None
+        """
+        for coord in self._points:
+            point = self.get_point(coord)
+            if point.value == point_value and point.is_visited == False:
+                return point
+
+    def get_point(self, coordinates):
+        """
+        Get element in Grid by its coordinates. Raises ValueError if coordinates are invalid.
+        :param coordinates: tuple(int,int)
+        :return: Point or raise ValueError

```

Comment:
Also happens to raise KeyError if the point is not in self.points yet. Probably not a problem but not evident from the docstring.